keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,652 @@
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validating admission webhook.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.
+
+### Notes/Constraints/Caveats (Optional)
+
+- In some cases, users really wan't to delete the objects that are considered to be protected in general.
+For example, immutable Secrets might need to be deleted while it is used by a Pod to update the value.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validating admission webhook will reject deletion requests for resources with non-empty `Liens`.
+
+`Liens` is defined as a slice of strings, like `Finalizers`.
+Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
+Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
+(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).
+
+To make multiple users and controllers share the `Liens` field properly, they should include their identity to the `Liens` string and they should only delete their own `Liens` strings.
+
+A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
+Also, how it can be consumed by Secret protection controller can be found, [here](TBD).
+
+### API Changes
+
+`Liens` is added to `ObjectMeta`.
+
+```
+Liens []string
+```
+
+### Other Design Considerations
```

Comment:
Missing: what happens if you delete a namespace and some objects within it have liens on them?

Reply:
Maybe it is also good to describe how lien works with `ownerReference`.

Differences of behaviors between "Foreground cascading deletion" and 
"Background cascading deletion" is expected to be a bit tricky. And behaviors of Namespace deletion will be like "Foreground cascading deletion" with `ownerReference.blockOwnerDeletion=true`.